### ISSUE TITLE
Loss function backward output kept in memory longer than expected

### ISSUE THREAD

**Issue by riccardofelluga** [Lightning-AI/lightning-thunder#1379]
## 🐛 Bug

The outputs from the log_softmax_backward is kept in memory till the start of the next forward whereas it should be released after use in the backward trace. 

### To Reproduce

```python
import torch
import thunder

def foo(a):
    t0 = torch.nn.functional.relu(a)
    t1 = torch.nn.functional.relu(t0)
    t2 = torch.nn.functional.relu(t1)
    t3 = torch.nn.functional.relu(t2)
    return t3

x = torch.randn((256, 256), requires_grad=True, device="cuda")
y = torch.randn_like(x)

jmlp = thunder.jit(foo, executors=[thunder.pytorch_executor], enable_saved_for_backward_recomputation=True)

torch.cuda.memory._record_memory_history()

for i in range(3):
    ins = x.detach().requires_grad_(True)
    out = jmlp(ins)
    out = torch.nn.functional.cross_entropy(out, y, ignore_index=-1)
    out.backward()
    out.item()
    del out

torch.cuda.memory._dump_snapshot("cross_entropy_test.pickle")
```

<img width="559" alt="repro_snap" src="https://github.com/user-attachments/assets/aae23422-1931-49f2-bcf0-9d8bfa585ee2">

**Comment by tfogal** [comment]
would this happen through thunderfx as well or is dynamo dealing with most of the saving tensors there?

**Comment by riccardofelluga** [comment]
I've seen the same behavior with ThunderFX by modifying the previous snippet:

```python
backend = ThunderCompiler(executors=[thunder.pytorch_executor], enable_saved_for_backward_recomputation=True)
jmlp = torch._dynamo.optimize(backend=backend)(foo)
```

I think this is expected as it happens between fwd and bwd thunder traces so after the ThunderFX pass

**Comment by kshitij12345** [comment]
I had a quick look at the backward traces with `enable_saved_for_backward_recomputation` being True and False

**Recomputation=True**: When recomputation is enabled, the earlier part of the backward trace is just recomputing and then post that we use the gradient from loss function (so it gets used later and deleted later, see `t65 = torch.where(t19, t10, 0)` in trace). Also,(from a quick glance) it seems that the recomputation performed at the earlier part is actually required in this order and can't be interleaved.

**Recomputation=False**: When recomputation is disabled, the first computation operation uses the gradient from loss and clears it. (See line `t65 = torch.where(t19, t10, 0)` in trace)

<details>

<summary> Backward Trace - enable_saved_for_backward_recomputation=True </summary>

```python
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  C0, _, = saved_for_backward
  clear_mutable_collection(saved_for_backward)
  del saved_for_backward
  t10, = cotangents
  clear_mutable_collection(cotangents)
  del cotangents
  a, = C0
  clear_mutable_collection(C0)
  del C0
  t8 = torch.gt(a, 0.0)  # t8: "cuda:0 b8[256, 256]"
    # t8 = ltorch.gt(a, 0.0)  # t8: "cuda:0 b8[256, 256]"
      # t8 = prims.gt(a, 0.0)  # t8: "cuda:0 b8[256, 256]"
  t1 = torch.where(t8, a, 0.0)  # t1: "cuda:0 f32[256, 256]"
    # t1 = ltorch.where(t8, a, 0.0)  # t1: "cuda:0 f32[256, 256]"
      # t1 = prims.where(t8, a, 0.0)  # t1: "cuda:0 f32[256, 256]"
  del a
  t11 = torch.gt(t1, 0.0)  # t11: "cuda:0 b8[256, 256]"
    # t11 = ltorch.gt(t1, 0.0)  # t11: "cuda:0 b8[256, 256]"
      # t11 = prims.gt(t1, 0.0)  # t11: "cuda:0 b8[256, 256]"
  t3 = torch.where(t11, t1, 0.0)  # t3: "cuda:0 f32[256, 256]"
    # t3 = ltorch.where(t11, t1, 0.0)  # t3: "cuda:0 f32[256, 256]"
      # t3 = prims.where(t11, t1, 0.0)  # t3: "cuda:0 f32[256, 256]"
  del t1
  t15 = torch.gt(t3, 0.0)  # t15: "cuda:0 b8[256, 256]"
    # t15 = ltorch.gt(t3, 0.0)  # t15: "cuda:0 b8[256, 256]"
      # t15 = prims.gt(t3, 0.0)  # t15: "cuda:0 b8[256, 256]"
  t5 = torch.where(t15, t3, 0.0)  # t5: "cuda:0 f32[256, 256]"
    # t5 = ltorch.where(t15, t3, 0.0)  # t5: "cuda:0 f32[256, 256]"
      # t5 = prims.where(t15, t3, 0.0)  # t5: "cuda:0 f32[256, 256]"
  del t3
  t19 = torch.gt(t5, 0.0)  # t19: "cuda:0 b8[256, 256]"
    # t19 = ltorch.gt(t5, 0.0)  # t19: "cuda:0 b8[256, 256]"
      # t19 = prims.gt(t5, 0.0)  # t19: "cuda:0 b8[256, 256]"
  del t5
  t65 = torch.where(t19, t10, 0)  # t65: "cuda:0 f32[256, 256]"
    # t65 = ltorch.where(t19, t10, 0)  # t65: "cuda:0 f32[256, 256]"
      # _ = prims.convert_element_type(0, float)
      # t65 = prims.where(t19, t10, 0.0)  # t65: "cuda:0 f32[256, 256]"
  del t19, t10
  t67 = torch.where(t15, t65, 0)  # t67: "cuda:0 f32[256, 256]"
    # t67 = ltorch.where(t15, t65, 0)  # t67: "cuda:0 f32[256, 256]"
      # _ = prims.convert_element_type(0, float)
      # t67 = prims.where(t15, t65, 0.0)  # t67: "cuda:0 f32[256, 256]"
  del t15, t65
  t69 = torch.where(t11, t67, 0)  # t69: "cuda:0 f32[256, 256]"
    # t69 = ltorch.where(t11, t67, 0)  # t69: "cuda:0 f32[256, 256]"
      # _ = prims.convert_element_type(0, float)
      # t69 = prims.where(t11, t67, 0.0)  # t69: "cuda:0 f32[256, 256]"
  del t11, t67
  t71 = torch.where(t8, t69, 0)  # t71: "cuda:0 f32[256, 256]"
    # t71 = ltorch.where(t8, t69, 0)  # t71: "cuda:0 f32[256, 256]"
      # _ = prims.convert_element_type(0, float)
      # t71 = prims.where(t8, t69, 0.0)  # t71: "cuda:0 f32[256, 256]"
  del t8, t69
  return (t71,)
```

</details>


<details>

<summary> Backward Trace - enable_saved_for_backward_recomputation=False </summary>

```python
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  C0, _, = saved_for_backward
  clear_mutable_collection(saved_for_backward)
  del saved_for_backward
  t10, = cotangents
  clear_mutable_collection(cotangents)
  del cotangents
  t11, t15, t19, t8, = C0
  clear_mutable_collection(C0)
  del C0
  t65 = torch.where(t19, t10, 0)  # t65: "cuda:0 f32[256, 256]"
    # t65 = ltorch.where(t19, t10, 0)  # t65: "cuda:0 f32[256, 256]"
      # _ = prims.convert_element_type(0, float)
      # t65 = prims.where(t19, t10, 0.0)  # t65: "cuda:0 f32[256, 256]"
  del t19, t10
  t67 = torch.where(t15, t65, 0)  # t67: "cuda:0 f32[256, 256]"
    # t67 = ltorch.where(t15, t65, 0)  # t67: "cuda:0 f32[256, 256]"
      # _ = prims.convert_element_type(0, float)
      # t67 = prims.where(t15, t65, 0.0)  # t67: "cuda:0 f32[256, 256]"
  del t15, t65
  t69 = torch.where(t11, t67, 0)  # t69: "cuda:0 f32[256, 256]"
    # t69 = ltorch.where(t11, t67, 0)  # t69: "cuda:0 f32[256, 256]"
      # _ = prims.convert_element_type(0, float)
      # t69 = prims.where(t11, t67, 0.0)  # t69: "cuda:0 f32[256, 256]"
  del t11, t67
  t71 = torch.where(t8, t69, 0)  # t71: "cuda:0 f32[256, 256]"
    # t71 = ltorch.where(t8, t69, 0)  # t71: "cuda:0 f32[256, 256]"
      # _ = prims.convert_element_type(0, float)
      # t71 = prims.where(t8, t69, 0.0)  # t71: "cuda:0 f32[256, 256]"
  del t8, t69
  return (t71,)
```

</details>

**Comment by riccardofelluga** [comment]
Tho it still seems to be keeping the tensor around even with `enable_saved_for_backward_reocomputation=False` (orange in the figure below)
![image](https://github.com/user-attachments/assets/02913304-9029-49b4-8bc0-61399ea6f9cc)

**Comment by kiya00** [comment]
Talked with Ivan, we found the tensor kept in memory longer is because the input of the backward function(`*args`) of  `torch.autograd.Function` is tuple, and it can not be freed until the end of the backward function:
https://github.com/Lightning-AI/lightning-thunder/blob/ea0159ca20b4109a84f572020b73220fb3c85328/thunder/executors/torch_autograd.py#L98

if we add the patch to print out the address of the `args`, it matches the address of the tensor kept in memory(the orange block) :
```
diff --git a/thunder/executors/torch_autograd.py b/thunder/executors/torch_autograd.py
index 5374b23a..216da944 100644
--- a/thunder/executors/torch_autograd.py
+++ b/thunder/executors/torch_autograd.py
@@ -113,6 +113,8 @@ class ThunderFunction(torch.autograd.Function):
         # reference to the saved tensors in the context
         ctx.maybe_clear_saved_tensors()  # Delete the reference to all saved tensors in the context
         grads = ctx.compiled_backward([saved_tensors_list, ctx.saved_other], args)
+        for arg in args:
+            print(hex(arg.data_ptr()))
 
         # Inside the compiled backward we must clear the saved_tensors_list
         assert not saved_tensors_list, "saved_tensors_list must be empty after calling compiled_backward"
```
<img width="488" alt="image" src="https://github.com/user-attachments/assets/050e2d0c-8112-4b44-a8ff-5d9ef3b67d8f">


So we can't release it early.

**Comment by kiya00** [comment]
As mentioned in https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670, the activations can be freed earlier by using `ctx.maybe_clear_saved_tensors()`, but `grad_output` can not be freed earlier.
here are the memory snapshot of nvfuser and torch.compile, we can observe the memory usage is similar
nvfuser:
<img width="1226" alt="image" src="https://github.com/user-attachments/assets/6484ad06-8f8e-4d2d-accf-2c6b9ad29019">
torch.compile:
<img width="560" alt="image" src="https://github.com/user-attachments/assets/4fcb6bff-770a-44bb-9a60-26c6026fc251">
So I think we can close this

**Comment by IvanYashchuk** [comment]
Thank you for the summary, Yan!
It would be great to find a way to hack through Python's behavior of keeping function arguments alive for the whole duration of the function even if a reference is deleted.

**Comment by t-vi** [comment]
> It would be great to find a way to hack through Python's behavior of keeping function arguments alive for the whole duration of the function even if a reference is deleted.

You can't at the Python level, if we wanted to be really evil, we could do `torch.utils.swap_tensors(grad_out, torch.empty((0,), device=grad_out.device))` to do it below the Python level. It would break anything using `grad_out` after us, though.

The other part is that we could just put the loss into the model.
